src/main/scala/com/microsoft/kusto/spark/utils/KustoClient.scala

```diff
@@ -5,33 +5,37 @@ import com.microsoft.kusto.spark.datasource._
 import com.microsoft.kusto.spark.utils.{KustoDataSourceUtils => KDSU}
 
 object KustoClient {
-  private[kusto] def getAdmin(authentication: KustoAuthentication, clusterAlias: String, isAdminCluster: Boolean = false): Client = {
-    val clusterUri = s"https://${if(isAdminCluster) "ingest"}$clusterAlias.kusto.windows.net"
-    val kcsb = getKcsbHelper(authentication,clusterUri)
+  def getAdmin(authentication: KustoAuthentication, clusterAlias: String, isIngestCluster: Boolean = false): Client = {
+    val clusterUri = s"https://${if(isIngestCluster) "ingest" else ""}$clusterAlias.kusto.windows.net"
+    val kcsb = getKcsb(authentication,clusterUri)
     kcsb.setClientVersionForTracing(KDSU.ClientName)
     ClientFactory.createClient(kcsb)
   }
 
-  private[kusto] def getIngest(authentication: KustoAuthentication, clusterAlias: String): IngestClient = {
-    val ingestKcsb = getKcsbHelper(authentication, s"https://ingest-$clusterAlias.kusto.windows.net")
+  def getAdmin(kcsb: ConnectionStringBuilder): Client = {
+    ClientFactory.createClient(kcsb)
+  }
+
+  def getIngest(authentication: KustoAuthentication, clusterAlias: String): IngestClient = {
+    val ingestKcsb = getKcsb(authentication, s"https://ingest-$clusterAlias.kusto.windows.net")
     ingestKcsb.setClientVersionForTracing(KDSU.ClientName)
     IngestClientFactory.createClient(ingestKcsb)
   }
 
-  private def getKcsbHelper(authentication: KustoAuthentication, clusterUri: String) ={
-    var kcsb: ConnectionStringBuilder = null
+  def getIngest(ingestKcsb: ConnectionStringBuilder): IngestClient = {
+    IngestClientFactory.createClient(ingestKcsb)
+  }
+
+  def getKcsb(authentication: KustoAuthentication, clusterUri: String): ConnectionStringBuilder = {
     authentication match {
+      case null => throw new MatchError("Can't create ConnectionStringBuilder with null authentication params")
       case app: AadApplicationAuthentication =>
-        kcsb = ConnectionStringBuilder.createWithAadApplicationCredentials(clusterUri, app.ID, app.password, app.authority)
+        ConnectionStringBuilder.createWithAadApplicationCredentials(clusterUri, app.ID, app.password, app.authority)
       case keyVaultParams: KeyVaultAuthentication =>
```

Comment:
Are KV get methods called only if the user provides KV credentials?
If sometimes called "opportunistically" consider wrapping with try-catch, and returning None on exception.

Reply:
only when user provieds